src/lib.rs

```diff
@@ -1722,6 +1722,20 @@ impl Build {
                         } else if target.contains("aarch64") {
                             cmd.args.push("--target=aarch64-unknown-windows-gnu".into())
                         }
+                    } else if target.ends_with("-freebsd") {
+                        // clang 13 on FreeBSD doesn't support a target triple without at least the
+                        // major os version number appended; e.g. use x86_64-unknown-freebsd13
+                        // instead of x86-64-unknown-freebsd
+                        let uname_r = String::from_utf8(
+                            std::process::Command::new("uname")
+                                .args(&["-r"])
+                                .output()
+                                .expect("Failed to execute uname!")
+                                .stdout,
+                        )
```

Comment:
cc-rs never makes the assumption that `target` is related `uname` output. However, it should be possible to make an assumption that if `self.get_host()` is the same as `target`, then clang would produce desired result **without** any `--target` option. I [for one] would even argue that it would be possible to generalize and make such assumption for **all** platforms, and replace the `else` below with `else if self.get_host()? != target`. One can even wonder if it would be appropriate to take it to the outer `if`...

Reply:
Yeah we should only do this in non-cross compilation situations. I also think we probably shouldn't use `uname -r` but `freebsd-version`, as is done here: https://github.com/rust-lang/libc/blob/106b57a796145c42f9a385a2a8cf764abbc66624/build.rs#LL160C46-L160C61

Reply:
@dot-asm Thanks for the `self.get_host()` hint; I wasn't sure how to find out if cross-compilation was intended or not. My preference was for excluding the triple altogether (this is the preference of the FreeBSD team, but read on), however I assumed that would not be palatable. I'm glad to hear it's something you're open to!

There is a very important caveat though: this only holds if the compiler is `clang(\+\+)?(-[0-9.]+)?` or `cc|c\+\+`. If `$CXX` or `$CC` was set and resulted in a different compiler being selected, we would need to specify and pass along the triple since there's no guarantee that the chosen compiler is native to the target (regardless of what `self.get_host()` returns, I believe). (This is at least the case if using a multiarch gcc install on Linux, at any rate.)

I'm not sure what we're supposed to do if the target triple *does* differ from the host triple. I would say we can just assume the triple has been set correctly in that case, but the problem is that cargo/rust triples differ from the llvm/clang ones (in that the version number is not appended). So the triple may very well be correct/valid *for rust* but insufficient for the targeted FreeBSD release. If we're explicitly in the cross-compilation branch, it doesn't make much sense to use the host `uname`/`freebsd-version` output. At the same time, I'm not aware of any mechanism that would allow us to introspect the actual intended target (which may not be available/live in the first place).

I feel like the answer there would require patching this logic elsewhere so that a user could invoke a build w/ target set to `x86_64-unknown-freebsd9` and cargo/rust would use the first part of that (without the trailing version number) but pass the whole thing to the compiler?

Reply:
> I feel like the answer there would require patching this logic elsewhere so that a user could invoke a build w/ target set to `x86_64-unknown-freebsd9`

I think we support things like setting the CC env var to something with flags, such as `CC="clang --target=x86_64-unknown-freebsd9"`.

Reply:
I didn't delve into the code deep enough to figure it out, but the question I see from here is whether or not `cmd.push_cc_arg("--target=xxx-yyy-zzz")` will defer to an existing CFLAGS/CPPFLAGS/CXXFLAGS `--target=aaa-bbb-ccc` value (rather than overwrite or append to it).

Reply:
> I'm glad to hear it's something you're open to!

Just in case, I'm just a cc-rs user, just like you. In other words what I'm open to doesn't have more weight than what you're open to :-)

> x86_64-unknown-freebsd9

Keep in mind that cc-rs is not responsible for linking. In other words if it's actually a significant/hard requirement, then wouldn't you have to find a way to make the rustc pass the same flag to the linker? But it's just straightforward `cc` even(*) in this case, right? This is basically why I suggested that it's possible to make the assumption that `self.gethost()? == target` warrants `--target`-less invocation. Because that's what rustc would normally do. Now, as for cross-compilation cases. Normally you'd have to explicitly specify the cross-linker, and it would be more appropriate to use **that** information to deduce the command-line for use in cc-rs. More appropriate than assuming this-or-that based on uname -r or freebsd-version ;-)

(*) "even" is a reference to Linux.

Reply:
I am treating compilation and linking as separate problems to solve; presumably if a linker was specified then either it is by default the correct linker for the target platform (i.e. correct target is hard-coded into its executable) or the required configuration variables have also been set in the environment to get it to work. If you rely on clang to execute the linker and you're not doing anything advanced, I'm pretty sure just executing `clang` w/ the correct `--target` and the default `--drive-mode` then it'll (try to) invoke the linker correctly. 

I'm also guessing basic cross-compilation and cross-linking is working or we'd have heard by now? (I have been wrong before!)

Reply:
> I am treating compilation and linking as separate problems to solve;

But they are not completely independent, things have to align. Not necessarily perfectly, but aligned nevertheless.

>  I'm pretty sure just executing clang w/ the correct --target and the default --drive-mode then it'll (try to) invoke the linker correctly.

And the point is that you have no control over --target passed to the linker. Not passing --target to cc simply mirrors the link stage and effectively aligns C compiler with Rust linker.

As for remark about FreeBSD people preferring target triplet over relying on --target-less invocation to target the current environment. Is it actually so that the option to not specify --target to match the current environment is about to disappear? Or is it rather that they want the option to target OS versions other than the current? I bet it's the latter. But in such case what do we achieve with the suggested modification? Do we give the user the option to choose? No, rather contrary actually, the user gets locked to the current system with no escape. Well, what do I know, maybe it is in fact the actual goal :-)